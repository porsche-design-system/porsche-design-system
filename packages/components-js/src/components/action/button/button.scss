@import '../../../../../../packages/scss-utils/index';

@mixin p-button-with-label() {
  padding: p-rem(11px) p-rem(15px) p-rem(11px) p-rem(39px);
}

@mixin p-button-without-label() {
  padding: 0;
}

@mixin p-button-label-with-label() {
  position: static;
  width: 100%;
  height: auto;
  margin: 0;
  padding: 0;
  white-space: normal;
  overflow: visible;
  border: 0;
  clip: auto;
  clip-path: none;
}

@mixin p-button-label-without-label() {
  position: absolute;
  width: 1px;
  height: 1px;
  margin: -1px;
  padding: 0;
  white-space: nowrap;
  overflow: hidden;
  border: 0;
  clip: rect(1px, 1px, 1px, 1px);
  clip-path: inset(50%);
}

@mixin p-button-icon-with-label() {
  left: p-rem(11px);
  top: p-rem(11px);
  transform: translateX(0) translateY(0);
}

@mixin p-button-icon-without-label() {
  left: 50%;
  top: 50%;
  transform: translateX(-50%) translateY(-50%);
}

:host {
  display: inline-flex;
  vertical-align: top;
}

.#{$p-global-class-prefix}button {
  $bem-block: &;
  display: flex;
  width: 100%;
  min-width: p-rem(48px);
  min-height: p-rem(48px);
  position: relative;
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  appearance: none;
  outline: none transparent;
  text-decoration: none;
  text-align: left;
  border-style: solid;
  border-width: 1px;
  transition: background-color $p-animation-hover-duration $p-animation-hover-bezier,
  border-color $p-animation-hover-duration $p-animation-hover-bezier,
  outline-color $p-animation-hover-duration $p-animation-hover-bezier,
  color $p-animation-hover-duration $p-animation-hover-bezier;
  cursor: pointer;

  &::-moz-focus-inner {
    border: 0; // fix for Firefox, avoid dotted outline in button on focus
  }

  &:focus {
    outline-width: 2px;
    outline-style: solid;
    outline-offset: 1px;
  }

  &:disabled,
  &[disabled] {
    cursor: not-allowed;
  }

  &--theme-light {
    &:focus {
      outline-color: $p-color-theme-light-state-focus;
    }
  }

  &--theme-dark {
    &:focus {
      outline-color: $p-color-theme-dark-state-focus;
    }
  }

  &--primary {
    &#{$bem-block}--theme-light {
      color: $p-color-theme-dark-default;
      background-color: $p-color-theme-light-brand;
      border-color: $p-color-theme-light-brand;
    }

    &#{$bem-block}--theme-dark {
      color: $p-color-theme-dark-default;
      background-color: $p-color-theme-dark-brand;
      border-color: $p-color-theme-dark-brand;
    }

    &:enabled,
    &:not([disabled]) {
      &:hover,
      &:active {
        &#{$bem-block}--theme-light {
          color: $p-color-theme-dark-default;
          background-color: darken($p-color-theme-light-state-hover, 12%);
          border-color: darken($p-color-theme-light-state-hover, 12%);
        }

        &#{$bem-block}--theme-dark {
          color: $p-color-theme-dark-default;
          background-color: darken($p-color-theme-dark-state-hover, 12%);
          border-color: darken($p-color-theme-dark-state-hover, 12%);
        }
      }
    }

    &:disabled,
    &[disabled] {
      &#{$bem-block}--theme-light {
        color: rgba($p-color-theme-dark-default, 0.55);
        background-color: $p-color-theme-light-state-disabled;
        border-color: $p-color-theme-light-state-disabled;
      }

      &#{$bem-block}--theme-dark {
        color: rgba($p-color-theme-dark-default, 0.55);
        background-color: $p-color-theme-dark-state-disabled;
        border-color: $p-color-theme-dark-state-disabled;
      }
    }
  }

  &--secondary {
    &#{$bem-block}--theme-light {
      color: $p-color-theme-dark-default;
      background-color: $p-color-theme-light-neutral-contrast-high;
      border-color: $p-color-theme-light-neutral-contrast-high;
    }

    &#{$bem-block}--theme-dark {
      color: $p-color-theme-light-default;
      background-color: $p-color-theme-dark-default;
      border-color: $p-color-theme-dark-default;
    }

    &:enabled,
    &:not([disabled]) {
      &:hover,
      &:active {
        &#{$bem-block}--theme-light {
          color: $p-color-theme-dark-default;
          background-color: darken($p-color-theme-light-neutral-contrast-high, 12%);
          border-color: darken($p-color-theme-light-neutral-contrast-high, 12%);
        }

        &#{$bem-block}--theme-dark {
          color: $p-color-theme-light-default;
          background-color: darken($p-color-theme-dark-default, 12%);
          border-color: darken($p-color-theme-dark-default, 12%);
        }
      }
    }

    &:disabled,
    &[disabled] {
      &#{$bem-block}--theme-light {
        color: rgba($p-color-theme-dark-default, 0.55);
        background-color: $p-color-theme-light-state-disabled;
        border-color: $p-color-theme-light-state-disabled;
      }

      &#{$bem-block}--theme-dark {
        color: rgba($p-color-theme-dark-default, 0.55);
        background-color: $p-color-theme-dark-state-disabled;
        border-color: $p-color-theme-dark-state-disabled;
      }
    }
  }

  &--tertiary {
    &#{$bem-block}--theme-light {
      color: $p-color-theme-light-default;
      background-color: transparent;
      border-color: $p-color-theme-light-neutral-contrast-high;
    }

    &#{$bem-block}--theme-dark {
      color: $p-color-theme-dark-default;
      background-color: transparent;
      border-color: $p-color-theme-dark-default;
    }

    &:enabled,
    &:not([disabled]) {
      &:hover,
      &:active {
        &#{$bem-block}--theme-light {
          color: $p-color-theme-dark-default;
          background-color: darken($p-color-theme-light-neutral-contrast-high, 12%);
          border-color: darken($p-color-theme-light-neutral-contrast-high, 12%);
        }

        &#{$bem-block}--theme-dark {
          color: $p-color-theme-light-default;
          background-color: $p-color-theme-dark-default;
          border-color: $p-color-theme-dark-default;
        }
      }
    }

    &:disabled,
    &[disabled] {
      &#{$bem-block}--theme-light {
        color: $p-color-theme-light-state-disabled;
        background-color: transparent;
        border-color: $p-color-theme-light-state-disabled;
      }

      &#{$bem-block}--theme-dark {
        color: $p-color-theme-dark-state-disabled;
        background-color: transparent;
        border-color: $p-color-theme-dark-state-disabled;
      }
    }
  }

  &--with-label {
    @include p-button-with-label;

    #{$bem-block}__label {
      @include p-button-label-with-label;
    }

    #{$bem-block}__icon {
      @include p-button-icon-with-label;
    }
  }

  &--without-label {
    @include p-button-without-label;

    #{$bem-block}__label {
      @include p-button-label-without-label;
    }

    #{$bem-block}__icon {
      @include p-button-icon-without-label;
    }
  }

  @include p-breakpoint("xs") {
    &--with-label-xs {
      @include p-button-with-label;

      #{$bem-block}__label {
        @include p-button-label-with-label;
      }

      #{$bem-block}__icon {
        @include p-button-icon-with-label;
      }
    }

    &--without-label-xs {
      @include p-button-without-label;

      #{$bem-block}__label {
        @include p-button-label-without-label;
      }

      #{$bem-block}__icon {
        @include p-button-icon-without-label;
      }
    }
  }

  @include p-breakpoint("s") {
    &--with-label-s {
      @include p-button-with-label;

      #{$bem-block}__label {
        @include p-button-label-with-label;
      }

      #{$bem-block}__icon {
        @include p-button-icon-with-label;
      }
    }

    &--without-label-s {
      @include p-button-without-label;

      #{$bem-block}__label {
        @include p-button-label-without-label;
      }

      #{$bem-block}__icon {
        @include p-button-icon-without-label;
      }
    }
  }

  @include p-breakpoint("m") {
    &--with-label-m {
      @include p-button-with-label;

      #{$bem-block}__label {
        @include p-button-label-with-label;
      }

      #{$bem-block}__icon {
        @include p-button-icon-with-label;
      }
    }

    &--without-label-m {
      @include p-button-without-label;

      #{$bem-block}__label {
        @include p-button-label-without-label;
      }

      #{$bem-block}__icon {
        @include p-button-icon-without-label;
      }
    }
  }

  @include p-breakpoint("l") {
    &--with-label-l {
      @include p-button-with-label;

      #{$bem-block}__label {
        @include p-button-label-with-label;
      }

      #{$bem-block}__icon {
        @include p-button-icon-with-label;
      }
    }

    &--without-label-l {
      @include p-button-without-label;

      #{$bem-block}__label {
        @include p-button-label-without-label;
      }

      #{$bem-block}__icon {
        @include p-button-icon-without-label;
      }
    }
  }

  @include p-breakpoint("xl") {
    &--with-label-xl {
      @include p-button-with-label;

      #{$bem-block}__label {
        @include p-button-label-with-label;
      }

      #{$bem-block}__icon {
        @include p-button-icon-with-label;
      }
    }

    &--without-label-xl {
      @include p-button-without-label;

      #{$bem-block}__label {
        @include p-button-label-without-label;
      }

      #{$bem-block}__iconÂ´ {
        @include p-button-icon-without-label;
      }
    }
  }

  &__label {
    display: block;
    box-sizing: border-box;
  }

  &__icon {
    position: absolute;
    width: p-rem(24px);
    height: p-rem(24px);
  }
}
