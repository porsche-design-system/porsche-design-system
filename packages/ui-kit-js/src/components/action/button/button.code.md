# Button

The `<p-button>` component is essential to perform events for forms or interactions. 

They can be used with or without a label but it's recommend to keep the label visible for better accessibility whenever possible. When used without a label, then it's best practice to provide a descriptive label text for screen readers.

## Variants

Choose between predefined styling variants.

### Primary

<Playground :themeable="true" :childElementLayout="{spacing: 'inline'}">
  <template v-slot={theme}>
    <p-button variant="primary" :theme="theme">Some label</p-button>
    <p-button variant="primary" disabled="true" :theme="theme">Some label</p-button>
    <p-button variant="primary" loading="true" :theme="theme">Some label</p-button>
    <br>
    <p-button variant="primary" hide-label="true" :theme="theme">Some label</p-button>
    <p-button variant="primary" hide-label="true" disabled="true" :theme="theme">Some label</p-button>
    <p-button variant="primary" hide-label="true" loading="true" :theme="theme">Some label</p-button>
  </template>
</Playground>

### Secondary (default)

<Playground :themeable="true" :childElementLayout="{spacing: 'inline'}">
  <template v-slot={theme}>
    <p-button :theme="theme">Some label</p-button>
    <p-button disabled="true" :theme="theme">Some label</p-button>
    <p-button loading="true" :theme="theme">Some label</p-button>
    <br>
    <p-button hide-label="true" :theme="theme">Some label</p-button>
    <p-button hide-label="true" disabled="true" :theme="theme">Some label</p-button>
    <p-button hide-label="true" loading="true" :theme="theme">Some label</p-button>
  </template>
</Playground>

### Tertiary

<Playground :themeable="true" :childElementLayout="{spacing: 'inline'}">
  <template v-slot={theme}>
    <p-button variant="tertiary" :theme="theme">Some label</p-button>
    <p-button variant="tertiary" disabled="true" :theme="theme">Some label</p-button>
    <p-button variant="tertiary" loading="true" :theme="theme">Some label</p-button>
    <br>
    <p-button variant="tertiary" hide-label="true" :theme="theme">Some label</p-button>
    <p-button variant="tertiary" hide-label="true" disabled="true" :theme="theme">Some label</p-button>
    <p-button variant="tertiary" hide-label="true" loading="true" :theme="theme">Some label</p-button>
  </template>
</Playground>

### Responsive

<Playground :themeable="true" :childElementLayout="{spacing: 'inline'}">
  <template v-slot={theme}>
    <p-button variant="primary" hide-label="{ base: true, s: false }" :theme="theme">Some label</p-button>
    <p-button variant="secondary" hide-label="{ base: true, m: false }" :theme="theme">Some label</p-button>
    <p-button variant="tertiary" hide-label="{ base: true, l: false }" :theme="theme">Some label</p-button>
  </template>
</Playground>

---

## Button with specific icon
If another icon needs to be implemented, just replace the default icon with another predefined icon. Per default, all icons are fetched from the Porsche Design System CDN. Just choose an icon name from the `icon` property. If you need to link to another icon hosted somewhere else, just set the whole icon path to the `iconSource` prop.

<Playground :themeable="true" :childElementLayout="{spacing: 'inline'}">
  <template v-slot={theme}>
    <p-button icon="delete" :theme="theme">Some label</p-button>
    <p-button :icon-source="require(`@/assets/web/icon-custom-kaixin.svg`)" hide-label="true" :theme="theme">Some label</p-button>
  </template>
</Playground>

---

## Bind events to the Button
You can use native `click`, `focus`, `focusin`, `blur` and `focusout` events on the button.

<Playground :themeable="true" :childElementLayout="{spacing: 'inline'}">
  <template v-slot={theme}>
    <p-button
        onclick="alert('click')"
        onfocus="console.log('focus')"
        onfocusin="console.log('focusin')"
        onblur="console.log('blur')"
        onfocusout="console.log('focusout')"
        :theme="theme"
    >Some label</p-button>
  </template>
</Playground>

---

## Remove Button from tab order
With setting the `tabbable` property to `false` you can remove the button from the tab order. For technical restrictions it's currently not possible to set an individual `tabindex` attribute.

<Playground :themeable="true" :childElementLayout="{spacing: 'inline'}">
  <template v-slot={theme}>
    <p-button tabbable="true" :theme="theme">Some label</p-button>
    <p-button tabbable="false" hide-label="true" :theme="theme">Some label</p-button>
  </template>
</Playground>
