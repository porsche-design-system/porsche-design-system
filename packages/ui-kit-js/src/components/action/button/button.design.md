# Button

A Button enables the user to execute an action or change the state of an application within the page. For an optimal user guidance and dedicated pursuit of business or sales goals, different types of buttons (primary, secondary, tertiary) can be used.

---

## Available variants

The Button is provided in 2 different variants:

- **Icon and Text** Default button size with 48 px height. The button length adapts to the length of the text label.
- **Icon only** The icon only contains an icon and no further text information, it is highly recommended to use it only in cases where the user is fully aware of the button function (this can be ensured e.g. by using an expressive icon or by logical composition with other components - just think of a form field followed by a Button Icon with submit function)

## Button types

### **1. P**rimary

<p-button variant="primary">Some label</p-button> 

Button with color fill in Porsche Red. This type is only to be used for one or two high-priority actions within the page, in order to give the user the best guidance possible.

### **2. S**econdary

<p-button>Some label</p-button>

Default button with monochrome color fill. To be used for all other stand-alone buttons that aren't high priority.

### **3. T**ertiary

<p-button variant="tertiary">Some label</p-button>

An "outline only" button version. In hierarchy the tertiary button is always subordinated to the filled view. Thus, it can be used in two different cases:

- In combination with a filled button (Primary or Secondary), for example in popular pairings such as "submit" (filled) and "cancel" (Tertiary).
- Stand-alone, when the priority of the action is lower compared to all other button actions within this page.

---

## Button states

All button types are available in the following states:

| STATE | DESCRIPTION |
|----|----|----|
| default | Default button state. |
| active / hover | In active or hover state, the background color changes slightly and gets either a bit lighter or darker. |
| disabled | Whenever the button function is not available, it is indicated by a greyed-out button color. |
| loading | To indicate the loading process when clicking on the text link, the icon is replaced by a small loading spinner. |
| focus | In focus state, the button is framed by a 2 px line in focus color. |

---

## Content

The Button is available for both Porsche Light and Dark Theme. It always contains the following sub-components:

### **Icon**
By default, an arrow right is set that should serve in most of all cases. It should be changed only if it is ensured that another symbol is more appropriate to support the text content, making it easier for the user to understand the function quickly. In addition, it should be ensured that the icon only variant is used for internationally known icons.

To ensure the alignment in all Porsche web appliccations we recommend to use dedicated, standardized icons for the following cases:

| INTERACTION | ICON | ADDITIONAL INFO | 
|----|----|----|
| Close/Cancel | <p-icon name="close" aria-label="Close icon"></p-icon> | Use a close icon for conclude content like modals and alerts (usally as an icon only) and a cancel icon for abandoning the process. | 
| Delete | <p-icon name="delete" aria-label="Delete icon"></p-icon> |  |
| Edit | <p-icon name="edit" aria-label="Edit icon"></p-icon> |  |
| Add | <p-icon name="plus" aria-label="Plusicon"></p-icon> |  |

### **Text label**
The text label is always set in copytext size. The length of the text label determines the size of the button. Multiline text is possible, while the content (both text and icon) are always positioned top left aligned within the button. It is recommended to keep the text label short and descriptive.

---

## Button position

By default the Button is to be positioned left-aligned within a module or a screen. Depending on content and user guidance, the position can be changed individually, e.g. placed right-aligned or at the end of a text (e.g. error notifications)

![Button position](./assets/button-position.png)

---

## Usability & interaction

### Clickability

The whole button area is clickable. The clickability is indicated by a specific hover state (slightly changing color).

### Button width

Even if there is no technical limit to the button width, you should always make sure that the button remains legible, even more so in multiline state. It is recommended to use max. 100 characters per line (equals approx. 700 px button width).

### Disabled state

All types of Button are available as disabled state. However, disabled states should be avoided whenever possible, as they always tend to disrupt the user and break the user flow. Keep in mind: “The best way [to] prevent user error is to make sure that the use cannot make errors in the first place (…).” (Donald A. Norman, 2002)

<div style="background:#F2F2F2; width:100%; margin-top: 64px; padding-top: 32px; padding-left: 42px; padding-bottom: 42px;">
    <p-headline variant="headline-3" tag="h3" style="margin-bottom: 24px;">Examples</p-headline>
    <img src="./assets/button-examples.png" alt=""/>
</div>

---

## Don'ts
  
![Example for uppercase text](./assets/button-don't.png)