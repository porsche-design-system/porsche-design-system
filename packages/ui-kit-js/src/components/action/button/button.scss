@import '../../../../../../packages/ui-kit-scss-utils/index';

@mixin p-button-with-label() {
  position: static;
  width: 100%;
  height: auto;
  margin: 0;
  padding: {
    top: rem(11px);
    right: rem(15px);
    bottom: rem(11px);
    left: rem(39px);
  }
  white-space: normal;
  overflow: visible;
  border: 0;
  clip: auto;
  clip-path: none;
}

@mixin p-button-without-label() {
  position: absolute;
  width: 1px;
  height: 1px;
  margin: -1px;
  padding: 0;
  white-space: nowrap;
  overflow: hidden;
  border: 0;
  clip: rect(1px, 1px, 1px, 1px);
  clip-path: inset(50%);
}

@mixin p-button-icon-with-label() {
  position: absolute;
  left: rem(11px);
  top: rem(11px);
  width: rem(24px);
  height: rem(24px);
  transform: translateX(0) translateY(0);
}

@mixin p-button-icon-without-label() {
  position: absolute;
  left: 50%;
  top: 50%;
  width: rem(36px);
  height: rem(36px);
  display: block;
  transform: translateX(-50%) translateY(-50%);
}

:host {
  display: inline-flex;
  vertical-align: top;
}

.#{$p-global-class-prefix}button {
  $bem-block: &;
  display: inline-flex;
  width: 100%;
  min-width: rem(48px);
  min-height: rem(48px);
  position: relative;
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  appearance: none;
  outline: none;
  text-decoration: none;
  border-style: solid;
  border-width: 1px;
  transition: background-color $p-animation-hover-duration $p-animation-hover-bezier,
    border-color $p-animation-hover-duration $p-animation-hover-bezier,
    outline-color $p-animation-hover-duration $p-animation-hover-bezier,
    color $p-animation-hover-duration $p-animation-hover-bezier;

  &::-moz-focus-inner {
    border: 0; // fix for Firefox, avoid dotted outline in button on focus
  }

  &--theme-light {
    outline-color: $p-color-porsche-light;
  }

  &--theme-dark {
    outline-color: $p-color-porsche-black;
  }

  &:focus {
    outline: 2px solid $p-color-state-focus;
    outline-offset: 1px;
  }

  &:disabled,
  &[disabled] {
    pointer-events: none;
  }

  &--primary {
    color: $p-color-porsche-light;
    background-color: $p-color-porsche-red;
    border-color: $p-color-porsche-red;

    &#{$bem-block}--theme-dark {
      color: $p-color-porsche-light;
      background-color: $p-color-porsche-red;
      border-color: $p-color-porsche-red;
    }

    &:enabled,
    &:not([disabled]) {
      &:hover,
      &:active {
        color: $p-color-porsche-light;
        background-color: $p-color-porsche-dark-red;
        border-color: $p-color-porsche-dark-red;

        &#{$bem-block}--theme-dark {
          color: $p-color-porsche-light;
          background-color: $p-color-porsche-dark-red;
          border-color: $p-color-porsche-dark-red;
        }
      }
    }

    &:disabled,
    &[disabled] {
      color: rgba($p-color-porsche-light, 0.55);
      background-color: $p-color-grey-disabled-on-light;
      border-color: $p-color-grey-disabled-on-light;

      &#{$bem-block}--theme-dark {
        color: rgba($p-color-porsche-light, 0.55);
        background-color: $p-color-grey-disabled-on-light;
        border-color: $p-color-grey-disabled-on-light;
      }
    }
  }

  &--secondary {
    color: $p-color-porsche-light;
    background-color: $p-color-neutral-grey-8;
    border-color: $p-color-neutral-grey-8;

    &--theme-dark {
      color: $p-color-porsche-black;
      background-color: $p-color-porsche-light;
      border-color: $p-color-porsche-light;
    }

    &:enabled,
    &:not([disabled]) {
      &:hover,
      &:active {
        color: $p-color-porsche-light;
        background-color: $p-color-neutral-grey-7;
        border-color: $p-color-neutral-grey-7;

        &#{$bem-block}--theme-dark {
          color: $p-color-porsche-black;
          background-color: $p-color-neutral-grey-1;
          border-color: $p-color-neutral-grey-1;
        }
      }
    }

    &:disabled,
    &[disabled] {
      color: rgba($p-color-porsche-light, 0.55);
      background-color: $p-color-grey-disabled-on-light;
      border-color: $p-color-grey-disabled-on-light;

      &#{$bem-block}--theme-dark {
        color: rgba($p-color-porsche-light, 0.55);
        background-color: $p-color-grey-disabled-on-dark;
        border-color: $p-color-grey-disabled-on-dark;
      }
    }
  }

  &--tertiary {
    color: $p-color-porsche-black;
    background-color: transparent;
    border-color: $p-color-neutral-grey-8;

    &#{$bem-block}--theme-dark {
      color: $p-color-porsche-light;
      background-color: transparent;
      border-color: $p-color-porsche-light;
    }

    &:enabled,
    &:not([disabled]) {
      &:hover,
      &:active {
        color: $p-color-porsche-light;
        background-color: $p-color-neutral-grey-8;
        border-color: $p-color-neutral-grey-8;

        &#{$bem-block}--theme-dark {
          color: $p-color-porsche-black;
          background-color: $p-color-porsche-light;
          border-color: $p-color-porsche-light;
        }
      }
    }

    &:disabled,
    &[disabled] {
      color: $p-color-grey-disabled-on-light;
      background-color: transparent;
      border-color: $p-color-grey-disabled-on-light;

      &#{$bem-block}--theme-dark {
        color: $p-color-grey-disabled-on-dark;
        background-color: transparent;
        border-color: $p-color-grey-disabled-on-dark;
      }
    }
  }

  &__label {
    display: block;
    box-sizing: border-box;

    &--show {
      @include p-button-with-label;

      & ~ #{$bem-block}__icon,
      & ~ #{$bem-block}__spinner {
        @include p-button-icon-with-label;
      }
    }

    &--hide {
      @include p-button-without-label;

      & ~ #{$bem-block}__icon,
      & ~ #{$bem-block}__spinner {
        @include p-button-icon-without-label;
      }
    }

    @include p-breakpoint("xs") {
      &--show-xs {
        @include p-button-with-label;

        & ~ #{$bem-block}__icon,
        & ~ #{$bem-block}__spinner {
          @include p-button-icon-with-label;
        }
      }

      &--hide-xs {
        @include p-button-without-label;

        & ~ #{$bem-block}__icon,
        & ~ #{$bem-block}__spinner {
          @include p-button-icon-without-label;
        }
      }
    }

    @include p-breakpoint("s") {
      &--show-s {
        @include p-button-with-label;

        & ~ #{$bem-block}__icon,
        & ~ #{$bem-block}__spinner {
          @include p-button-icon-with-label;
        }
      }

      &--hide-s {
        @include p-button-without-label;

        & ~ #{$bem-block}__icon,
        & ~ #{$bem-block}__spinner {
          @include p-button-icon-without-label;
        }
      }
    }

    @include p-breakpoint("m") {
      &--show-m {
        @include p-button-with-label;

        & ~ #{$bem-block}__icon,
        & ~ #{$bem-block}__spinner {
          @include p-button-icon-with-label;
        }
      }

      &--hide-m {
        @include p-button-without-label;

        & ~ #{$bem-block}__icon,
        & ~ #{$bem-block}__spinner {
          @include p-button-icon-without-label;
        }
      }
    }

    @include p-breakpoint("l") {
      &--show-l {
        @include p-button-with-label;

        & ~ #{$bem-block}__icon,
        & ~ #{$bem-block}__spinner {
          @include p-button-icon-with-label;
        }
      }

      &--hide-l {
        @include p-button-without-label;

        & ~ #{$bem-block}__icon,
        & ~ #{$bem-block}__spinner {
          @include p-button-icon-without-label;
        }
      }
    }

    @include p-breakpoint("xl") {
      &--show-xl {
        @include p-button-with-label;

        & ~ #{$bem-block}__icon,
        & ~ #{$bem-block}__spinner {
          @include p-button-icon-with-label;
        }
      }

      &--hide-xl {
        @include p-button-without-label;

        & ~ #{$bem-block}__icon,
        & ~ #{$bem-block}__spinner {
          @include p-button-icon-without-label;
        }
      }
    }
  }
}
