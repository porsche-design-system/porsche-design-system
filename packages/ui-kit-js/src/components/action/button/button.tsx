import { JSX, Component, Prop, State, h, Element, Listen } from '@stencil/core';
import cx from 'classnames';
import { prefix, hasShadowDom } from '../../../utils';
import { IconName } from '../../icon/icon/icon-name';
import { improveFocusHandlingForCustomElement, preventNativeTabIndex } from '../../../utils/focusHandling';
import {ButtonType, Theme} from '../../../types';

@Component({
  tag: 'p-button',
  styleUrl: 'button.scss',
  shadow: true
})
export class Button {
  @Element() public element!: HTMLElement;

  /**
   * Check native tabindex to ensure that it doesn't get set on the host element
   * @internal
   */
  @Prop({
    mutable: true,
    attribute: 'tabindex'
  })
  public nativeTabindex?: number = -1;

  /** To remove the element from tab order. */
  @Prop() public tabbable?: boolean = true;

  /** Specifies the type of the button. */
  @Prop() public type?: ButtonType = 'button';

  /** A visually hidden label text to improve accessibility which describes the function behind the button. */
  @Prop() public allyLabel?: string = undefined;

  /** Disables the button. No events will be triggered while disabled state is active. */
  @Prop() public disabled?: boolean = false;

  /** Disables the button and shows a loading indicator. No events will be triggered while loading state is active. */
  @Prop() public loading?: boolean = false;

  /** The style variant of the button. */
  @Prop() public variant?: 'primary' | 'secondary' | 'tertiary' = 'secondary';

  /** The icon shown. */
  @Prop() public icon?: IconName = 'plus';

  /** A custom URL path to a custom icon. */
  @Prop() public iconSource?: string = undefined;

  /** Adapts the button color when used on dark background. */
  @Prop() public theme?: Theme = 'light';

  /** Show or hide label */
  @Prop() public hideLabel?: boolean = false;

  @State() public isSlotDefined?: boolean;

  private slots: NodeListOf<ChildNode>;

  public componentWillLoad() {
    this.slots  = this.element.childNodes;
    // this.slots.length !== 0 ? this.isSlotDefined = true : this.isSlotDefined = false;
    this.slots.length === 0 || this.hideLabel && this.slots.length !== 0 ? this.isSlotDefined = false : this.isSlotDefined = true;
  }

  public componentDidLoad() {
    improveFocusHandlingForCustomElement(this.element);
    // console.log(this.element);
    // console.log('123', this.element.shadowRoot.querySelectorAll('slot'));
  }

  public componentWillUpdate() {
    this.slots.length === 0 || this.hideLabel && this.slots.length !== 0 ? this.isSlotDefined = false : this.isSlotDefined = true;
  }

  public render(): JSX.Element {
    preventNativeTabIndex(this);

    const buttonClasses = cx(
      prefix('button'),
      this.variant !== 'secondary' && prefix(`button--${this.variant}`),
      this.theme !== 'light' && prefix('button--theme-dark')
    );
    const iconClasses = prefix('button__icon');
    const spinnerClasses = prefix('button__spinner');
    const labelClasses = prefix('button__label');

    return (
      <button
        class={buttonClasses}
        type={this.type}
        disabled={this.disabled || this.loading}
        aria-label={this.allyLabel}
        tabindex={this.tabbable ? 0 : -1}
      >
        {this.isSlotDefined ? (
          <p-text tag='span' color='inherit' class={labelClasses}>
            <slot/>
          </p-text>
        ) : null}
        {this.loading ? (
          <p-spinner class={spinnerClasses} size='x-small' theme={this.useInvertedLoader()} />
        ) : (
          <p-icon class={iconClasses} size='inherit' name={this.icon} source={this.iconSource} />
        )}
      </button>
    );
  }


  /**
   * IE11 workaround to fix the event target
   * of click events (which normally shadow dom
   * takes care of)
   */
  @Listen('click', { capture: true })
  public fixEventTarget(event: MouseEvent): void {
    if (event.target !== this.element) {
      event.stopPropagation();
      event.preventDefault();
      this.element.click();
    }
  }

  @Listen('click')
  public onClick(event: MouseEvent): void {
    if (this.type === 'submit' && hasShadowDom(this.element)) {
      // Why? That's why: https://www.hjorthhansen.dev/shadow-dom-and-forms/
      const form = this.element.closest('form');
      if (form) {
        /**
         * we've to wait if someone calls preventDefault on the event
         * then we shouldn't submit the form
         */
        window.setTimeout(() => {
          if(!event.defaultPrevented) {
            const fakeButton = document.createElement('button');
            fakeButton.type = this.type;
            fakeButton.style.display = 'none';
            form.appendChild(fakeButton);
            fakeButton.addEventListener('click', (fakeButtonEvent) => {
              fakeButtonEvent.stopPropagation();
            });
            fakeButton.click();
            fakeButton.remove();
          }
        }, 1);
      }
    }
  }

  private useInvertedLoader(): Theme {
    return this.variant !== 'tertiary' || this.theme === 'dark' ? 'dark' : 'light';
  }

}
