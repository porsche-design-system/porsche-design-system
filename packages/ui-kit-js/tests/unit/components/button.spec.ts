import { newSpecPage } from '@stencil/core/testing';
import { Button } from '../../../src/components/action/button/button';

describe('Component <p-button>', () => {

  it('should build', () => {
    expect(new Button()).toBeTruthy();
  });

  it('should render correctly in default mode with shadow dom', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<p-button>Some label</p-button>`,
    });
    expect(page.root.shadowRoot).toBeTruthy();
    expect(page.root.querySelector('.p-button')).toBeFalsy();
    expect(page.root.shadowRoot.querySelector('.p-button')).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('should render with type of submit', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<p-button type="submit">Some label</p-button>`,
    });
    expect(page.root.shadowRoot).toBeTruthy();
    expect(page.root.shadowRoot.querySelector('button')).toEqualAttribute('type','submit');
  });

  it('should render with an aria label text', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<p-button ally-label="Text string"></p-button>`,
    });
    expect(page.root.shadowRoot).toBeTruthy();
    expect(page.root.shadowRoot.querySelector('button')).toEqualAttribute('aria-label','Text string');
  });

  it('should have a disabled prop in button mode', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<div></div>`
    });

    const component = page.doc.createElement('p-button');

    (component as any).disabled = true;
    page.root.appendChild(component);
    await page.waitForChanges();

    expect(page.rootInstance.disabled).toBe(true);
  });

});

