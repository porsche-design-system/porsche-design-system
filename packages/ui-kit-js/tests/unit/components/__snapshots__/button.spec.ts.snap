// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <p-button> should render correctly in default mode with shadow dom 1`] = `
<p-button>
  <mock:shadow-root>
    <button class="p-button" tabindex="0" type="button">
      <p-text class="p-button__label p-button__label--show" color="inherit" tag="span">
        <slot></slot>
      </p-text>
      <p-icon class="p-button__icon" name="plus" size="inherit"></p-icon>
    </button>
  </mock:shadow-root>
  Some label
</p-button>
`;
